ansible-galaxy - increase page size and add retry decorator for throttling

s-hertel opened the pull request:
##### SUMMARY
Fixes #74191

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/galaxy/api.py

s-hertel committed:
Add backoff wrapper for Galaxy API throttling

s-hertel committed:
Get available collection versions with page_size=100

Check for 'page' in the query to load from and refresh the cache since the query may be empty except for page_size

Co-authored-by: Matt Martz <matt@sivel.net>

s-hertel commented on lib/ansible/galaxy/api.py:
```diff
@@ -46,6 +47,28 @@ def wrapped(*args, **kwargs):
+            except GalaxyError as e:
+                # Ideally, this would handle rate limit errors and not unknown errors.
+                # TODO: Check if v3 also uses 520 error code for rate limit errors, or if that's a v2 quirk.
+                if e.http_code not in (520, 429):
```
Today I've also been experiencing sporadic 502 errors.

webknjaz replied:
I'd argue that these should be configurable too.

Also, it'd be great to move this to a module-global constant like:
```python
from ansible.module_utils.six.moves import http_client


RETRIABLE_HTTP_CODES = {
    http_client.TOO_MANY_REQUESTS,
    520,  # internal rate limit code?
}
```

It's best to put magic numbers into constants so that when used, they'd add more meaning to the code.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -46,6 +47,27 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def rate_limit_backoff(function):
```
Any particular reason we can't use/vendor https://github.com/litl/backoff? That lib has a fantastic API and we'd not have to reimplement its slick UX.

Also, we've encountered other places where the backoff is used and is sometimes poorly implemented. I recall some modules needed it in the past. So it'd make sense to have it made available in a generic fashion as opposed to another module-specific implementation.

s-hertel replied:
I guess we could vendor it? It's such a minimal change that I didn't think it was necessary. I thought we may want to backport, and if so, we shouldn't add a new dep (that's also why things aren't user-configurable yet as that's feature-ish).

sivel replied:
Generally speaking, I'd like us to be very cautious on adding any new deps, and I'm not a huge fan of vendoring.  That lib also isn't available in a number of distros that I've just checked.

bcoca replied:
we already have these type functions in module_utils/api.py as decorators

s-hertel replied:
@bcoca we don't have one with a specifiable backoff strategy or one that can handle specific exception types or error codes. Should I move this to common code? I debated over where to put it.

bcoca replied:
yes, this is useful for other things, so i would put it under module_utils, even expand existing functions if it makes sense

webknjaz replied:
@sivel FWIW if we were to use backoff, we'd have to vendor it to make this available to modules on target hosts especially because it's useful for modules talking to services over API.

@s-hertel yep, specifically because this is useful to modules it should to under `module_utils`.

P.S. I understand that this change seems small and only proposed vendoring backoff because _I saw_ similar problems solved at least in two modules in the past by reinventing the wheel. If you take into account those other cases, the generic solution gets bigger, and then it may make sense to just vendor a third-party lib and expose a nice API as a bonus + maybe add a small wrapper just for this module if that's necessary.

webknjaz replied:
So out of curiosity, I did a quick GH repo search and here's my findings:
* exponential backoff in modules:
  * https://github.com/ansible/ansible/blob/4c5ce5a1a9e79a845aff4978cfeb72a0d4ecf7d6/lib/ansible/modules/apt_repository.py#L645
  * https://github.com/ansible/ansible/blob/2bff120db6bf96dd0085c6ce1b29f3999b872db4/lib/ansible/modules/apt.py#L1189
* in action plugins:
  * https://github.com/ansible/ansible/blob/a51a6f4a259b45593c3f803737c6d5d847258a83/lib/ansible/plugins/action/reboot.py#L302
* in connection plugins:
  * https://github.com/ansible/ansible/blob/96905120698e3118d8bafaee5ebe8f83d2bbd607/lib/ansible/plugins/connection/psrp.py#L859 (delegated to Jordan's lib — `pypsrp`)
* in tests:
  * https://github.com/ansible/ansible/blob/a077bca5d5866603cdc4a2a13aef8f416860aacb/test/support/integration/plugins/modules/ec2_vpc_subnet.py#L259 (`AWSRetry` — looks like a post-migration leftover, not actually used)
* in `module_utils`:
  * https://github.com/ansible/ansible/blob/a077bca5d5866603cdc4a2a13aef8f416860aacb/test/support/integration/plugins/module_utils/cloud.py#L48
  * https://github.com/ansible/ansible/blob/a51a6f4a259b45593c3f803737c6d5d847258a83/lib/ansible/module_utils/api.py#L94 (not exponential but linear backoff)
* and there's even snippets in docs showing how to implement this in YAML playbooks: 
  *  https://github.com/ansible/ansible/blob/ccd9a992cf7b7b057cfd9a2d06c69c44b7094696/docs/docsite/rst/user_guide/complex_data_manipulation.rst#L28 (not related to our problem I guess)
* also, this very module has exponential backoff already baked-in in one place too:
  * https://github.com/ansible/ansible/blob/00bd0b893d5d21de040b53032c466707bacb3b93/lib/ansible/galaxy/api.py#L689

s-hertel replied:
The first two links (apt and apt_repository) will be simple to replace with a generic wrapper.

`reboot` has some debug messages during the retry process that would overcomplicate a common retry decorator, in my opinion.

The exponential backoff in the `ansible/lib/ansible/galaxy/api.py` is actually handling two things - retrying errors and waiting until the import is complete. Those are coupled together and will need some refactoring to use a common retry decorator.

I'm consolidating code with the `retry` function in `lib/ansible/module_utils/api.py` (not extending it, as it's doing some not-so-generic-things). I'll fix `apt` and `apt_repository` in a follow-up PR.

Everything else looks unrelated. The docs are great for things that don't implement retries. The stuff in `ansible/test/support` needs to stay as-it-is. `psrp` is a false positive since the option is for the python library, and not a case of Ansible reinventing the wheel.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -46,6 +47,27 @@ def wrapped(*args, **kwargs):
+            try:
+                return function(*args, **kwargs)
+            except GalaxyError as e:
+                # Galaxy can't always return a helpful error code because cloud.redhat.com masks it
```
I'm not sure what this comment tries to explain. How does this problem manifest itself in our runtime?

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -781,7 +805,7 @@ def get_collection_versions(self, namespace, name):
             pagination_path = ['next']
 
-        versions_url = _urljoin(self.api_server, api_path, 'collections', namespace, name, 'versions', '/')
+        versions_url = _urljoin(self.api_server, api_path, 'collections', namespace, name, 'versions', '/?page_size=100')
```
Could you move this magic value into an external constant?

sivel commented:
This will also need adjusted for galaxy_ng,  which uses `offset`/`limit` instead of `page`/`page_size`.

s-hertel committed:
Fix unit tests for v2

webknjaz commented on test/units/galaxy/test_api.py:
```diff
@@ -802,9 +803,9 @@ def test_get_collection_versions(api_version, token_type, token_ins, response, m
         {
             'count': 6,
-            'next': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=2',
+            'next': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=2?page_size=100',
```
```suggestion
            'next': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=2&page_size=100',
```

webknjaz commented on test/units/galaxy/test_api.py:
```diff
@@ -817,7 +818,7 @@ def test_get_collection_versions(api_version, token_type, token_ins, response, m
         {
             'count': 6,
-            'next': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=3',
+            'next': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=3?page_size=100',
```
```suggestion
            'next': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=3&page_size=100',
```

webknjaz commented on test/units/galaxy/test_api.py:
```diff
@@ -833,7 +834,7 @@ def test_get_collection_versions(api_version, token_type, token_ins, response, m
             'count': 6,
             'next': None,
-            'previous': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=2',
+            'previous': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=2?page_size=100',
```
```suggestion
            'previous': 'https://galaxy.server.com/api/v2/collections/namespace/collection/versions/?page=2&page_size=100',
```

webknjaz commented on test/units/galaxy/test_api.py:
```diff
@@ -916,12 +918,14 @@ def test_get_collection_versions_pagination(api_version, token_type, token_ins,
 
     assert mock_open.call_count == 3
+
+    page_query = '?limit=100' if api_version == 'v3' else '?page_size=100'
```
```suggestion
    page_query = '&limit=100' if api_version == 'v3' else '&page_size=100'
```

webknjaz commented on test/units/galaxy/test_api.py:
```diff
@@ -791,9 +791,10 @@ def test_get_collection_versions(api_version, token_type, token_ins, response, m
     actual = api.get_collection_versions('namespace', 'collection')
     assert actual == [u'1.0.0', u'1.0.1']
 
+    page_query = '?limit=100' if api_version == 'v3' else '?page_size=100'
```
```suggestion
    page_query = '&limit=100' if api_version == 'v3' else '&page_size=100'
```

s-hertel committed:
Make the retry decorator generic and move it to module_utils

Update the existing retry function in module_utils/api.py to use the generic retry decorator

Fix v3 pagination

ci_complete

s-hertel commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
+    if retries <= 1:
+        return _retry_limit_exceeded(retries)
```
This is a bit messy. You can't have `raise Exception(...)` in a lambda, so I made another function just to do that. I'm of half a mind to leave `retry` as it was rather than make it use the common function since it was doing some not-so-generic things.

s-hertel replied:
Removed this from the PR.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -36,6 +38,11 @@
 
 display = Display()
 _CACHE_LOCK = threading.Lock()
+COLLECTION_PAGE_SIZE = '100'
```
Should this be a string? It's already automatically `str()`ingified

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -36,6 +38,11 @@
+COLLECTION_PAGE_SIZE = '100'
+RETRY_HTTP_ERROR_CODES = [  # TODO: Allow user-configuration
+    429,  # Too Many Requests
+    520,  # Galaxy rate limit error code
```
Maybe add a comment that this is actually Cloudflare's code?

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -46,6 +53,15 @@ def wrapped(*args, **kwargs):
+    # Note: cloud.redhat.com masks rate limit errors with 403 (Forbidden) error codes.
+    # Since 403 could reflect the actual problem (such as an expired token), we should
+    # not retry by default.
+    if isinstance(exception, GalaxyError) and exception.http_code in RETRY_HTTP_ERROR_CODES:
+        return True
+    return False
```
The if-block seems a litte excessive, could just do
```suggestion
    return isinstance(exception, GalaxyError) and exception.http_code in RETRY_HTTP_ERROR_CODES
```

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -46,6 +53,15 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def rate_limit_exception(exception=None):
```
Any particular reason to have a default? Also, this looks like a test-function / predicate. So it's better to prefix it with `is_` to communicate the action it performs.

s-hertel replied:
For a while (while trying to refactor `reboot`) this function was actually called as `except retry_condition():` and the exception was retrieved and returned from `sys`.

webknjaz replied:
I think adopting the interface outlined @ https://github.com/ansible/ansible/pull/74240/files#r615180154 should accommodate this case.

s-hertel replied:
It looks like it could accommodate it in the same way as this implementation could - which is, not well. `giveup` (or in this implementation, `retry_condition`) would need to accept the delay in addition to the exception (and args/kwargs additional to those used by the function being retried) and it would do the extra stuff as a side-effect when `giveup` would return False (or in this implementation, when retry_condition return True). I'm still poking at `reboot` (maybe I should add a `do_on_retry` callable parameter?), but it seems like an excessive special case.

s-hertel replied:
I added a do_on_retry optional parameter. Here's what the reboot module could look like refactored to use the common code: https://gist.github.com/s-hertel/b87bad21adea00e4a91fd68a307e5fc6

webknjaz replied:
That looks nice (although I have comments on the complexity/nesting but that's unrelated)

s-hertel replied:
:smile: I'll keep that in mind and try to simplify when I open the PR.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -309,10 +325,12 @@ def available_api_versions(self):
         # Calling g_connect will populate self._available_api_versions
         return self._available_api_versions
 
+    @retry_wrapper(jittered_backoff_generator(retries=6, delay=2, max_delay=40), rate_limit_exception)
```
It feels like it's best to use explicit kwargs — when I read `rate_limit_exception` I automatically think that it's an instance of some exception object while in fact reading its definition it turns out to be a callable.
And even for the first argument, it'd be useful to know what you set by passing what I assume is an iterator.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -781,7 +799,10 @@ def get_collection_versions(self, namespace, name):
             pagination_path = ['next']
 
-        versions_url = _urljoin(self.api_server, api_path, 'collections', namespace, name, 'versions', '/')
+        if 'v3' in self.available_api_versions:
+            versions_url = _urljoin(self.api_server, api_path, 'collections', namespace, name, 'versions', '/?limit=%s' % COLLECTION_PAGE_SIZE)
+        else:
+            versions_url = _urljoin(self.api_server, api_path, 'collections', namespace, name, 'versions', '/?page_size=%s' % COLLECTION_PAGE_SIZE)
```
I think that copy-pasting the invocation can make this prone to human error during refactoring because one would have to update two code branches w/o making typos. Mind making just the query arg name conditional?
```suggestion
        page_size_name = 'limit' if 'v3' in self.available_api_versions else 'page_size'
        versions_url = _urljoin(self.api_server, api_path, 'collections', namespace, name, 'versions', '/?%s=%s' % (page_size_name, COLLECTION_PAGE_SIZE))
```

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
-    return wrapper
+
+def retry(retries=None, retry_pause=1):
+    """Retry decorator.
```
PEP 257 nit
```suggestion
    """Retry decorator.

```

s-hertel committed:
Apply suggestions from code review

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+    have been reached. If retries is None, the function is a no-op.
+    """
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
```
This is clever but it's broken my brain until I parsed it :)

The general maintainability advice (most linters actually enforce) is to pass lambdas as arguments when calling other functions (like passing predicates to filters and stuff).

Please move this to a named function. At least expand the outer function so it could be used to communicate it purpose.
```python
def noop(_func):
    return lambda *_args, **_kwargs: None
```

`return noop` reads better, doesn't it?

s-hertel replied:
I'm leaning towards just removing the changes to `retry`. It can be refactored to use the common function, but I just feel like it doesn't gain much of anything.

webknjaz replied:
Yeah, that'd work too. I'm just trying to give you a fresh-look perspective here...

s-hertel replied:
Removed the refactor.

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+    if retries <= 1:
+        return _retry_limit_exceeded(retries)
+    try:
+        backoff_delays = [retry_pause for delay in range(0, retries)]
+        return retry_wrapper(backoff_iterator=backoff_delays, retry_condition=lambda x: True)
```
In one place you pass a lazy iterator (`jittered_backoff_generator()`) and here you pass a pre-populated list. Maybe unify the style and pass a generator expression?
```suggestion
        return retry_wrapper(backoff_iterator=range(retries), retry_condition=lambda x: True)
```

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+def _retry_limit_exceeded(retries):
+    def function_wrapper(function):
+        def raise_error(*args, **kwargs):
+            raise Exception("Retry limit exceeded: %d" % retries)
```
This exception is too broad. It'd be useful to declare a specific error:
```suggestion
            raise RetryError("Retry limit exceeded: %d" % retries)
```

s-hertel replied:
I'd rather not modify the existing code unnecessarily, but I'll probably remove the `retry` changes anyway, and hence delete this function to revisit another time in a PR that isn't intended for backport.

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+    try:
+        backoff_delays = [retry_pause for delay in range(0, retries)]
+        return retry_wrapper(backoff_iterator=backoff_delays, retry_condition=lambda x: True)
+    except Exception:
```
This turns any error into a broad rate limit exception, since we control the emitted exception we can catch just it and any other exception wouldn't be silently shadowed.

s-hertel replied:
I'm not sure what you mean or what you'd like me to change.

webknjaz replied:
I'd like `except RetryError` I think.

s-hertel replied:
~Okay, good idea :+1:~

Oh wait. This is the `retry` function and I'm not modifying behavior here. This can be done in another PR perhaps, but I don't think it should be backported as something could be relying on `Exception`.

s-hertel replied:
I've *refactored* this function, but not intended to change behavior. Anyway, I'm going to just remove the refactor. No need to backport this for sure, and I don't think the refactor is worth much anyway :smile:

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+
+def jittered_backoff_generator(retries=10, delay=3, max_delay=60):
+    for retry in range(0, retries):
+        yield random.randint(0, min(max_delay, delay * 2 ** retry))
```
If I'm reading this right, it's a full jitter as defined by https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/.

Maybe add a note about that?

s-hertel replied:
:+1: Yes, it is.

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+        raise Exception("Retry limit exceeded: %d" % retries)
+
+
+def jittered_backoff_generator(retries=10, delay=3, max_delay=60):
```
This probably shouldn't be a noun and say what it does right in the name:
```suggestion
def emit_jittered_backoff_delays(retries=10, delay=3, max_delay=60):
```
Communicating more specific information in a function name makes it easier to reason about the calling code.

s-hertel replied:
"backoff generator" means what it does to me, but I can rename it.

webknjaz replied:
There's a convention to call things that do stuff with verbs hinting what the action is while things that just store some state (like constants) go by nouns.

Just compare:
```python
do_something = 42
def do_something():
    print(42)
def a_car():
    drive()
a_car = Car()
a_car.drive()
```

Ref: https://www.youtube.com/watch?v=YklKUuDpX5c

s-hertel replied:
`generator` is descriptive to me because I know it will yield values. 'emit' is an unknown, I'd actually have to look at the code to know what that means. But I don't have a problem with renaming this.

webknjaz replied:
How about `generate_jittered_backoff()` then?

s-hertel replied:
I like that :+1:

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
+                try:
+                    return function(*args, **kwargs)
+                except Exception as e:
+                    if retry_condition(e):
+                        time.sleep(delay)
+                    else:
+                        raise e
```
Let's simplify this a bit:
```suggestion
                    if not retry_condition(e):
                        raise

                    time.sleep(delay)

```

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -91,26 +92,58 @@ def ratelimited(*args, **kwargs):
+        yield random.randint(0, min(max_delay, delay * 2 ** retry))
+
+
+def retry_wrapper(backoff_iterator, retry_condition):
```
```suggestion
def retry_with_delays(backoff_iterator, retry_condition):
```
(wrapper is an implementation detail, it's probably more informative to say that it adds some delays)

Also, thinking about `backoff_iterator` — maybe it's more like a strategy? Not sure tho.

s-hertel replied:
backoff_iterator and strategy make me think of a delay generator, not a decorator doing the retries

`retry_with_delays_and_condition`?

webknjaz replied:
Yeah, probably. It feels like you're getting closer to implementing the API of https://github.com/litl/backoff — so maybe just copy their invocation signature (at least the args that make sense for us) and go from there? It's pretty neat.

s-hertel replied:
I don't see the point to incorporating a 3rd party lib unnecessarily. I do think it's neat, but this was loosely based on the old lib/ansible/module_utils/ec2.py stuff. That stuff has worked for ages with a wide variety of AWS services all off doing their own thing, why reinvent the wheel.

webknjaz replied:
I'm not saying that we should use `backoff`, I say that we should adopt some of its API definitions (as in the types of arguments `@backoff()` accepts — some, not all).

s-hertel replied:
Can you give me a link with the signature you're thinking of? I don't really know what you're referring to in particular, and I don't see it from a quick scan of the repo.

webknjaz replied:
I was thinking along the lines of https://github.com/litl/backoff/blob/e20982dc36be213465a1e26864d8e9792fd3af92/backoff/_decorator.py#L18 / https://github.com/litl/backoff/blob/e20982dc36be213465a1e26864d8e9792fd3af92/backoff/_decorator.py#L105.

I think the useful interface parts for us are:
```python
@backoff.on_predicate(
    wait_gen,  # iterator, emitter of delays
    predicate,  # a callable accepting the decorated function return value and deciding whether to retry on not (returns bool)
    max_tries,
    max_time,
    jitter,  # delay transformer callable, converts values produced by wait_gen
)
```
and
```python
@backoff.on_exception(
    wait_gen,  # iterator, emitter of delays
    exception,  # exception triggering retry
    max_tries,
    max_time,
    jitter,  # delay transformer callable, converts values produced by wait_gen
    giveup,  # optional callable analyzing the exception additionally to the `exception` argument, can judge whether to retry or reraise
)
```

I like that this is composed of decoupled components — `wait_gen` and `giveup` are separate rather than coupled, retries/time are not tied to jitter and wait_gen. And the way to cleanly provide a retriable exception (could be a tuple I guess) is just brilliant — it allows the reader of the calling code to easily reason about what it does.

The helper names could be `retry_on_exception` and `retry_if`. Although, maybe we only need one variant of this interface, not both.

s-hertel replied:
* wait_gen
* max_tries
* max_time
* jitter

Taking these seems less generic than taking an iterable of delays because it's implying an implementation. For example, if this were to be used in the `reboot` module, the delays would be yielded until a max total duration (which seems like a nice thing to support from a user-perspective), not a pre-determined number of attempts.

* predicate

This is nice, basically you can have a retry decorator double as a waiter. There's nothing that would use it yet, but maybe I should add it as an argument now so we don't need to add it as a kwarg later for backwards compatibility? It will make it easy to handle the existing galaxy/api.py retries when that's refactored.

* exception

What's the benefit to taking a specific classes when the conditional can check `isinstance`?

* giveup

The current name seems fitting since this is operating on the assumption that no exceptions will be retried by default instead of backoff's assumption that all will be retried.

s-hertel replied:
I added a 'retry_on_success' parameter similar to predicate. Let me know what you think - I know the signature is not the same but functionality-wise, this feels a little more flexible to me.

webknjaz replied:
> What's the benefit to taking a specific classes when the conditional can check `isinstance`?

Just for sugar but yeah, that's not mandatory.

> I added a 'retry_on_success' parameter similar to predicate. Let me know what you think - I know the signature is not the same but functionality-wise, this feels a little more flexible to me.

Yep, the signature doesn't need to be exactly the same. I just thought it'd demonstrate better what I meant.

I feel like `retry_on_success` reads like it accepts a boolean (rather than a callable) that will cause a retry whenever the decorated $thing succeeds. So I'd probably choose a different name.

s-hertel committed:
Remove retry refactor to use common code

s-hertel committed:
semantics

s-hertel committed:
Oops

s-hertel committed:
Add two optional parameters to retry_with_delays_and_condition: a callable that runs before the delay and a callable to retry when the decorated function successfully returns

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -322,7 +341,7 @@ def _call_galaxy(self, url, args=None, headers=None, method=None, auth_required=
                 valid = datetime.datetime.utcnow() < expires
 
-            if valid and not url_info.query:
+            if valid and not ('page' in query or 'offset' in query):
```
The second condition became complex enough that id turn it into a named expression:
```suggestion
            is_paginated_url = 'page' in query or 'offset' in query
            if valid and not is_paginated_url:
```
WDYT?

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,53 @@ def retried(*args, **kwargs):
+
+
+def generate_jittered_backoff(retries=10, delay=3, max_delay=60):
+    """The "Full Jitter" backoff strategy described here https://www.awsarchitectureblog.com/2015/03/backoff.html
```
Let's make this PEP 257 compliant:
```suggestion
    """The "Full Jitter" backoff strategy.

    Ref: https://www.awsarchitectureblog.com/2015/03/backoff.html
```

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,53 @@ def retried(*args, **kwargs):
+    :param backoff_iterator: An iterable of delays in seconds.
+    :param retry_condition: A callable that takes an exception of the decorated function and decides whether to retry or not (returns a bool).
+    :param do_on_retry: An optional callable that takes an exception/None and the current delay and runs before the sleep (returns None).
+    :param retry_on_success: An optional callable that accepts the decorated function return value and decides whether or not to retry (returns a bool).
```
I think I'd call it `retry_or_not` or `should_retry`. This way it wouldn't only refer to a positive boolean outcome.

s-hertel replied:
Do you think `retry_or_not`/`should_retry` would sound like it should run regardless of whether the function call was successful though? This callable should run if the decorated function doesn't raise an exception.

webknjaz replied:
That's an interesting observation. Maybe use `should_retry_on_error` then? Thinking about it, this is probably why backoff called their args `giveup` and `predicate` — the first one clearly relates to an error caught (and an expectation that something failed and there was no successful result which is a typical pythonic idiom) and the latter pretty obviously inspects the return value in a scenario when the errors aren't communicated via exceptions.

s-hertel replied:
I've renamed both of the retry conditions to `should_retry_error` and `should_retry_result`, lmkwyt

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,53 @@ def retried(*args, **kwargs):
+                        raise
+                    error = e
+                else:
+                    if not callable(retry_on_success) or not retry_on_success(result):
+                        return result
```
Provided that it's renamed to `should_retry` as suggested above (https://github.com/ansible/ansible/pull/74240/files#r617759827), I think this should look like
```suggestion
                    if not should_retry(result):
                        return result
```
This would be possible if we define a default value for this argument outside of `function_wrapper`:
```python
if should_retry is None:
    should_retry = retry_always
```
Also, a nice side-effect would be that defining `retry_always` on the module level makes it importable (and it could be reused with a nice name in `reboot`):
```python
def retry_always(_):
    return True
```

s-hertel replied:
done

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,53 @@ def retried(*args, **kwargs):
+                    if not callable(retry_on_success) or not retry_on_success(result):
+                        return result
+
+                if callable(do_on_retry):
+                    do_on_retry(error, delay)
```
Same here:
```suggestion
                do_on_retry(error, delay)
```
and
```python
if do_on_retry is None:
    do_on_retry = lambda *_, *__: None
```
outside `function_wrapper`

s-hertel committed:
Apply suggestions from code review

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -342,7 +362,7 @@ def _call_galaxy(self, url, args=None, headers=None, method=None, auth_required=
                 return res
 
-            elif not url_info.query:
+            elif not ('page' in query or 'offset' in query):
```
I must've explicitly mentioned that the var could be reused here too:
```suggestion
            elif not is_paginated_url:
```

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,55 @@ def retried(*args, **kwargs):
     return wrapper
+
+
+def generate_jittered_backoff(retries=10, delay=3, max_delay=60):
```
Could `delay` be `delay_base`? The explanation below says that it's initial but the reality is that it's used on each iteration, not just the first one...

Also `max_delay` sounds more like a `delay_threshold` maybe?

WDYT?

s-hertel replied:
renamed those :+1:

s-hertel committed:
Rename and set defaults for some parameters

s-hertel committed:
Fix callable assignment

webknjaz commented on test/units/module_utils/test_api.py:
```diff
@@ -46,3 +46,11 @@ def login_database():
+
+        @retry()
+        def login_database():
+            raise Exception("Should not execute")
```
I believe this could be
```suggestion
            assert False, "Should not execute"
```

webknjaz replied:
Or
```suggestion
            pytest.fail("Should not execute")
```

webknjaz reviewed (approved):
LGTM. Are you planning to add more tests for the retry utils?

s-hertel committed:
Make do_on_retry accept the exception or result

Add unit tests

s-hertel commented:
@webknjaz I added some unit tests.

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,69 @@ def retried(*args, **kwargs):
+                try:
+                    result = function(*args, **kwargs)
+                except Exception as e:
+                    if not should_retry_error(e):
+                        raise
+                    error = e
+                else:
+                    if not should_retry_result(result):
+                        return result
+
+                do_on_retry(delay=delay, exception=error, result=result)
```
Looking at this once again, it looks like `do_on_retry()` got too many args and scope creep along with that.

Basically, `should_retry_error()` and `should_retry_result()` can have any side-effects you want already. Do we _really_ need an extra way to do this?

One could do 
```python
def do_common_thing():
    time.sleep(5)

def should_retry_error(error):
    do_common_thing()
    return False

def should_retry_result(result):
    do_common_thing()
    return False

@retry_with_delays_and_condition(
    [...],
    should_retry_error=should_retry_error,
    should_retry_result=should_retry_result
)
```

s-hertel replied:
That doesn't accomplish the same thing. Those functions would need to accept the delay. Even if those functions can technically do other stuff, I'd consider it bad code organization because it would be a side-effect in those functions.

The reason `do_on_retry` exists is because I was trying to handle the `reboot` refactor in the future, and it uses the current delay to make helpful debug information during retry attempts. I decided to include `result` because I can see a scenario when you might want to make a `waiter` that also includes debug information (only including the exception seemed arbitrary). A significant part of this PR could be considered 'scope creep' because it was intended to solve an individual thing and turned into a generic utility, most of which will not be utilized in the backport, for which this is intended.

webknjaz replied:
I see. It still feels a bit foreign but I guess we can't do anything about it. Maybe an extra person review would be useful.

s-hertel replied:
I decided to strip out everything that has no reason to be backported. With commit history, it will be easy to add those kwargs back when the other places that need those changes are refactored.

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,69 @@ def retried(*args, **kwargs):
+                result = None
+
+                try:
+                    result = function(*args, **kwargs)
```
nit: since this is used in two places, it'd be easier to maintain it if we bake in the args.
How do you feel about doing
```python
call_retriable_thing = functools.partial(function, *args, **kwargs)
```
at the top of `run_function` and then just doing `call_retriable_thing()` and on return.

webknjaz commented on test/units/module_utils/test_api.py:
```diff
@@ -46,3 +46,119 @@ def login_database():
+        self.counter = 0
+
+        @retry_with_delays_and_condition(backoff_iterator=[])
+        def login_database():
+            self.counter += 1
+
```
Maybe we could avoid relying a shared instance attribute?
```suggestion
        def login_database():
            login_database.counter += 1
        login_database.counter = 0

```

webknjaz replied:
Another (and possibly cleaner) way to do this is to use `pytest-mock`:

```python
class SpiedOnTarget:
    @staticmethod
    def do_a_thing():
        return None

def test_empty_retry_iterator(mocker):
    decorate_with_no_retries = retry_with_delays_and_condition(backoff_iterator=[])
    spyable = mocker.spy(SpiedOnTarget, do_a_thing)
    invoke_retriable = decorate_with_no_retries(SpiedOnTarget.do_a_thing)

    invoke_retriable()
    assert spyable.call_count == 0
```

s-hertel replied:
`mocker.spy` is great. I used your first suggestion though, since the retry-able functions that need an initial failure are actually using the call count.

webknjaz replied:
Oh, the other tests?

webknjaz commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +115,69 @@ def retried(*args, **kwargs):
+    if should_retry_result is None:
+        should_retry_result = retry_never
+
+    def function_wrapper(function):
```
I overlooked this small detail — the current implementation doesn't copy over the underlying function's metadata:
```suggestion
    @functools.wraps(function)  # FIXME: use wrapt once it's in
    def function_wrapper(function):
```
This is necessary for the discovery in REPL and docs generation tools to pick up the proper `__doc__` and the argspec.

Also, there's talk, https://wrapt.rtfd.io is going to become a runtime dependency for the sake of object proxies feature in 2.12. So in the future, it's probably best to use it instead (AFAIK it does a better job taking care of the decorator corner cases).

s-hertel replied:
Great catch, thanks.

s-hertel committed:
Remove premature generalization and the corresponding unit tests

Preserve the decorated function's metadata with functools.wraps

Some other requested changes

webknjaz commented on test/units/module_utils/test_api.py:
```diff
@@ -46,3 +45,53 @@ def login_database():
+                raise Exception("Error")
+
+        login_database.counter = 0
+        with pytest.raises(Exception):
```
Always use raises with `match=` or you'll catch false positives. Especially because almost any exception is a subclass of `Exception`.
```suggestion
        with pytest.raises(Exception, match='Error'):
```

webknjaz replied:
Honestly, I'd change it to some `class CustomTestException(Exception): ...` to avoid any weird corner cases.

webknjaz commented on test/units/module_utils/test_api.py:
```diff
@@ -46,3 +45,53 @@ def login_database():
+    def test_retry_exception(self):
+        @retry_with_delays_and_condition(
+            backoff_iterator=[1],
+            should_retry_error=lambda x: isinstance(x, Exception),
```
It'd be interesting to see a test case for something inheriting a `BaseException` too.

s-hertel committed:
stuff

s-hertel committed:
and

samdoran commented on lib/ansible/module_utils/api.py:
```diff
@@ -114,3 +116,51 @@ def retried(*args, **kwargs):
     return wrapper
+
+
+def generate_jittered_backoff(retries=10, delay_base=3, delay_threshold=60):
```
This is great. Much better than the exponential backoff I implemented for `reboot`.

s-hertel replied:
Thanks. After this is merged I'm going to refactor the other places that have implemented retries to use common code (apt, apt_repository, reboot, and the GalaxyAPI import process).

samdoran reviewed (approved):
This is really well done. 👏

s-hertel commented:
Thanks for the reviews @webknjaz and @samdoran!
